python/sglang/srt/disaggregation/conn.py

```diff
@@ -364,21 +486,102 @@ def failure_exception(self):
+            port = self.available_ports.pop()
+            return Port
+
+    def release_port(self, port):
+        with self.lock:
+            if port < self.start_port or port > self.end_port:
+                raise Exception(f"Port {port} out of range")
+            self.available_ports.add(port)
+
+    def get_available_ports(self):
+        with self.lock:
+            return len(self.available_ports)
+
+
 class KVBootstrapServer:
     def __init__(self, port: int):
+        self.route_port = port
         self.port = port
         self.app = web.Application()
         self.store = dict()
         self.lock = asyncio.Lock()
         self._setup_routes()
 
+        self.context = zmq.Context()
+
+        # Route socket to communicate with Dealer which is prefill or decode
+        self.router_socket = self.context.socket(zmq.ROUTER)
+        self.router_socket.bind(f"tcp://*:{self.route_port}")
+
+        self.prefill_engine_rank = None
+        self.prefill_tp_size = None
+
         # Start bootstrap server
         self.thread = threading.Thread(target=self._run_server, daemon=True)
-        self.run()
-
-    def run(self):
         self.thread.start()
 
+        # Start listen clients thread
+        self.zmq_thread = threading.Thread(target=self._listen_clients, daemon=True)
+        self.zmq_thread.start()
+
+        # Decode port pool, hard-coded the port range for now
+        self.decode_port_pool = PortPool(35000, 35500)
+
+        self.tp_sizes = {}
+
+    def _listen_dealers(self):
+        while True:
+            client_id, message = self.router_socket.recv_multipart()
+
+            request_data = pickle.loads(message)
+            if request_data["type"] == "STARTUP":
+                identity = request_data["identity"]
+                role = request_data["role"]
+                tp_size = request_data["tp_size"]
+
+            role = role.decode("ascii")
+            engine_rank = int(engine_rank.decode("ascii"))
+            tp_size = int(tp_size.decode("ascii"))
+            # TODO(yuan-luo): construct a topology map for a prefill rank
+            # to locate a decode rank
+            self.tp_sizes[role] = int(tp_size)
+            if role == "Prefill":
+                self._handle_prefill(engine_rank, tp_size)
+            elif role == "Decode":
+                self._handle_decode()
+
+    def _handle_prefill(self, engine_rank, tp_size):
+        """Handle Prefill message"""
+        self.prefill_engine_rank = engine_rank
+        self.prefill_tp_size = tp_size
+
+    def _handle_decode(self):
+        """Handle Decode message"""
+        if self.prefill_engine_rank is None or self.prefill_tp_size is None:
+            print("Error: Metadata not yet received from Prefill.")
+            self.router_socket.send_multipart([identity, b"Error: Metadata not ready"])
+        else:
+            self.router_socket.send_multipart(
+                [
+                    "Decode".encode("ascii"),
```

Comment:
ROUTER socket need explicitly send `identity` as the first part of a message, as this part will be handled inside the ZMQ library for routing the message to the correct DEALER socket and removed when the message is sent back to the DEALER socket

Reply:
Resolved. Added client_id.